### PR TITLE
Python: Make the type tracking implementation shareable

### PR DESCRIPTION
**Pull request by tausbn**
cc: @nickrolfe @aibaars @hvitved 

This is a first stab at making type tracking shareable between languages. The broad strokes are

- `Attribute` has been renamed to `Content` in order to be language-agnostic.
- Language-specific dependencies have been moved into the `TypeTrackerPrivate.qll` file.

The `Private` file must expose a few predicates and classes:
- `getPossibleContentName`, which returns the strings we wish to consider as content,
- `typePreservingStep`, which should hold for any predecessor and successor nodes that should have the same type. (It's possible this should be called `levelStep` instead.)
- `callStep`, which should link up arguments with parameters
- `returnStep` which should link up return values with call nodes.
- `basicStoreStep` which relates the value that is being written to an attribute/property/field with a given name, to ~the _earliest_ source of~ the object being written to.
- `basicLoadStep` which goes the other way, relating an attribute/property/field lookup to its result.
- `Node` which should just be the class of nodes in the shared dataflow library, and
- `LocalSourceNode` which should contain the nodes that represent the earliest points of local flow.
- `Boolean`, which wraps the built-in `boolean` type.

In the case of Python, `TypeTrackerPrivate.qll` has to be written in a slightly torturous way because `TypeTracker` is already defined in our version of the data-flow library. This should not be necessary for languages that do not have this problem.

### CODE REVIEW

File: python/ql/src/experimental/typetracking/TypeTracker.qll
Comment (RasmusWL) [inline python/ql/src/experimental/typetracking/TypeTracker.qll:5]:
```suggestion
/** 
 * Any string that may appear as the name of a piece of content. This will usually include things like:
 * - Attribute names (in Python)
 * - Property names (in JavaScript)
 */
```

I think that having the explanation for what "content" is here in the shared implementation, is also useful

**Review by RasmusWL — COMMENTED** [review]
I'd like to better understand what "content" is. Is it only restricted to attributes/properties? (but just using a neutral name), or would it also be allowed to model list elements with something like `[*]`? (just an example)

Besides that, just one minor comment I would like to get clarified :blush: otherwise LGTM :+1:

**Comment by tausbn** [comment]
I agree that a better description of what content can and cannot be should be added here. (The present implementation was written in somewhat of a hurry.)

I'm also tempted to move things around a bit more. Currently, the somewhat magical `basicStoreStep` has the magic (i.e. jumping back to the local source of the object being accessed) happen inside the language-specific bit, but it might make more sense to share the magic in the common implementation, and simply require that `basicStoreStep` looks for local accesses of content, but _doesn't_ account for local flow in itself.

On the other hand, it might be the case that such magic is undesirable in some cases, which would speak against including it by default.

File: python/ql/src/experimental/typetracking/TypeTracker.qll
Comment (intrigus-lgtm) [inline python/ql/src/experimental/typetracking/TypeTracker.qll:100]:
```suggestion
 * ```ql
```

File: python/ql/src/experimental/typetracking/TypeTracker.qll
Comment (intrigus-lgtm) [inline python/ql/src/experimental/typetracking/TypeTracker.qll:261]:
```suggestion
 * ```ql
```

**Comment by RasmusWL** [comment]
> I'm also tempted to move things around a bit more. Currently, the somewhat magical `basicStoreStep` has the magic (i.e. jumping back to the local source of the object being accessed) happen inside the language-specific bit, but it might make more sense to share the magic in the common implementation, and simply require that `basicStoreStep` looks for local accesses of content, but _doesn't_ account for local flow in itself.
> 
> On the other hand, it might be the case that such magic is undesirable in some cases, which would speak against including it by default.

I think having the magic shared by default sounds like the most promising solution. If it ever turns out to be a problem, we can just move it out to the specific language implementations.

**Comment by tausbn** [comment]
Having addressed the review comments, this is now in a much nicer state, but I think it could be nicer still. I would like to add some documentation at the top of `TypeTracker.qll` describing what type trackers are and how they work (and in particular what their shortcomings are).

File: python/ql/src/experimental/typetracking/TypeTracker.qll
Comment (nickrolfe) [inline python/ql/src/experimental/typetracking/TypeTracker.qll:230]: `StepSummary::step` restricts the type of `nodeTo` to `LocalSourceNode`. Shouldn't the type here also be `LocalSourceNode`, or should `StepSummary::step` change?

Comment (tausbn) [reply python/ql/src/experimental/typetracking/TypeTracker.qll:230]: You're right, it should. I must have somehow overlooked this when I created this version. :+1:

Comment (nickrolfe) [reply python/ql/src/experimental/typetracking/TypeTracker.qll:230]: You mean the type on `StepSummary::step` should just be `Node`? That would fix missing flow that I've seen.

Comment (tausbn) [reply python/ql/src/experimental/typetracking/TypeTracker.qll:230]: No, they should both be `LocalSourceNode` (I somehow overlooked that your question contained an alternative). If you're missing flow, we should probably have a chat about that.

**Comment by RasmusWL** [comment]
@tausbn if @nickrolfe is already using this, should we prioritize actually using this "shared" version in Python?

**Comment by tausbn** [comment]
> @tausbn if @nickrolfe is already using this, should we prioritize actually using this shared version in Python?

Naturally. But we can't switch over until we can change _our_ `StepSummary::step` predicate to use `LocalSourceNode`s everywhere. 

Incidentally, have you noticed a pattern in my recent PRs? :wink:

File: python/ql/src/experimental/typetracking/TypeTracker.qll
Comment (hvitved) [inline python/ql/src/experimental/typetracking/TypeTracker.qll:3]: For other uses of pyrameterized modules, we have typically called this `TypeTrackerSpecific.qll`.

File: python/ql/src/experimental/typetracking/TypeTrackerPrivate.qll
Comment (hvitved) [inline python/ql/src/experimental/typetracking/TypeTrackerPrivate.qll:1]: I suggest a file doc along the lines of
```ql
/**
 * Provides Python-specific definitions for use in the type tracker library.
 */
```

**Review by hvitved — COMMENTED** [review]
Some trivial comments

**Review by aibaars — COMMENTED** [review]
This looks really good to me. Thanks for pulling this out and making it shareable.

File: python/ql/src/experimental/typetracking/TypeTracker.qll
Comment (tausbn) [reply python/ql/src/experimental/typetracking/TypeTracker.qll]: Done.

File: python/ql/src/experimental/typetracking/TypeTrackerPrivate.qll
Comment (tausbn) [reply python/ql/src/experimental/typetracking/TypeTrackerPrivate.qll]: Done.

File: python/ql/src/experimental/typetracking/TypeTracker.qll
Comment (tausbn) [reply python/ql/src/experimental/typetracking/TypeTracker.qll:6]: I think "content" is more appropriate, precisely because it is not limited to just attributes/fields/properties. See for instance the  JavaScript modelling of [array accesses](https://github.com/github/codeql/blob/main/javascript/ql/src/semmle/javascript/Arrays.qll#L165). It's perhaps not immediate from the context, but this `loadStep` is used to create an `AdditionalTypeTrackingStep`, and in this case the "content" (or "property" in the JS libs) is `arrayElement()` which ultimately points to a [string value](https://github.com/github/codeql/blob/main/javascript/ql/src/semmle/javascript/dataflow/Configuration.qll#L797) of `$arrayElement$` -- an unlikely name for an object attribute. 
(For what I assume are legacy reasons, the JS libs call these "pseudo-properties", but rather than picking a too-specific name and then subsequently making it more generic, we have the chance to make it generic from the get-go. :slightly_smiling_face:)

Finally, the name "content" is also used in much the same way in [the shared dataflow library](https://github.com/github/codeql/blob/main/csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPublic.qll#L192), so I think it has merit to reuse this name (though if the two notions end up end up being too different, this might cause some confusion).

**Comment by tausbn** [comment]
I have updated the PR with some improvements based on my recent experience with getting rid of most of the custom type trackers in the Python models (in favour of API graphs instead). The commit messages should explain what's going on (but do ask questions if there is something unclear).

I'm also not quite done with this PR. I recently had some epiphanies about some of the design decisions in this library, and I would like to document these insights before getting it merged. I do not currently plan on changing the code further, however.

File: python/ql/src/experimental/typetracking/TypeTracker.qll
Comment (aibaars) [reply python/ql/src/experimental/typetracking/TypeTracker.qll:6]: I was afraid that the only reason for the word "content" was to be language agnostic . I agree with using the same terminology as the shared dataflow library and if it is truly meant to be more generic then just field/property/attribute then "content" makes sense. Thanks for the explanation.

**Review by aibaars — APPROVED** [review]

**Comment by aibaars** [comment]
@tausbn I'll merge this as-is for now. You probably want to move it out of experimental and add more documentation in a follow-up pull request. Thanks for pulling out the shareable parts of the type-tracking library.

**Pull request merged**
